### ISSUE TITLE
kernel keeps dying after a few seconds

### ISSUE THREAD

**Issue by kirk86** [Calysto/matlab_kernel#24]
Environment and software packages:
----------------------------------------------
OS: Ubuntu 14.04
 ipython (4.0.0)
jupyter-client (4.1.1)
pymatbridge (0.5.2)
metakernel (0.11.3)
matlab-kernel (0.6.9)

MATLAB_EXECUTABLE=/usr/local/MATLAB/.../.... etc

I can/could simply fire up matlab from the linux command prompt just fine.
I added the above MATLAB_EXECUTABLE path variable  to my .bashrc in case that could possibly help alleviate the problem but nothing has actually changed. Still matlab kernels just die in a matter of couple of seconds just right after I have started a new matlab notebook.

Actual Error message:
----------------------------
KernelRestarter: restarting kernel (1/5)
WARNING:root:kernel 907a3e8a-33fb-451c-b980-81306574de94 restarted
/usr/local/lib/python2.7/dist-packages/IPython/nbformat.py:13: ShimWarning: The `IPython.nbformat` package has been deprecated. You should import from nbformat instead.
  "You should import from nbformat instead.", ShimWarning)
[IPKernelApp] ERROR | Can't load '/usr/local/lib/python2.7/dist-packages/metakernel/magics/python_magic.py': error: unqualified exec is not allowed in function 'eval' it contains a nested function with free variables (python_magic.py, line 63)

Thanks!

**Comment by dsblank** [comment]
Interesting. The problem appears to be:
```
/usr/local/lib/python2.7/dist-packages/metakernel/magics/python_magic.py
```
What version, exactly, of Python are you using?

**Comment by kirk86** [comment]
Hi, I am using Python 2.7.6.

**Comment by dsblank** [comment]
We are testing with Python 2.7.9:

https://travis-ci.org/Calysto/metakernel/jobs/86858330#L80

Could that me the difference?

**Comment by kirk86** [comment]
If I go for Python 2.7.10 will it be ok or is it strictly related to 2.7.9.

**Comment by dsblank** [comment]
On Sun, Oct 25, 2015 at 12:51 PM, kirk86 <notifications@github.com> wrote:

> If I go for Python 2.7.10 will it be ok or is it strictly related to 2.7.9.
>

We will make sure that MetaKernel will work with all future versions of
Python 2.7.x and 3.x.

**Comment by kirk86** [comment]
So I updated to Python 2.7.10 but again I receive the same error messages as above.
I tried to uninstall matlab_kernel, metakernel and pymatbridge via pip and install again via pip.
Installation give the following warnings:

"writing list of installed files to '/tmp/pip-gz1slj-record/install-record.txt'
    /usr/local/lib/python2.7/dist-packages/IPython/kernel/__init__.py:13: ShimWarning: The `IPython.kernel` package has been deprecated. You should import from ipykernel or jupyter_client instead.
      "You should import from ipykernel or jupyter_client instead.", ShimWarning)
    Installing kernel spec
"

 When I fire up the notebook again I receive the same error and kernel dies after a few seconds.

Something else that I noticed is the following. When I test the kernel via the following command 'ipython qtconsole --kernel=matlab_kernel' on qtconsole I receive the same error: 

"ipython qtconsole --kernel=matlab_kernel
/usr/local/lib/python2.7/dist-packages/IPython/nbformat.py:13: ShimWarning: The `IPython.nbformat` package has been deprecated. You should import from nbformat instead.
  "You should import from nbformat instead.", ShimWarning)
[IPKernelApp] ERROR | Can't load '/usr/local/lib/python2.7/dist-packages/metakernel/magics/python_magic.py': error: unqualified exec is not allowed in function 'eval' it contains a nested function with free variables (python_magic.py, line 63)"

After a while and although the qtconsolse seems unresponsive I actually have a matlab kernel runnings and I can type actual commands in the qtconsole but I can't get the same behavior in the jupyter notebook.

**Comment by blink1073** [comment]
That is a weird one.  @kirk86, can you see if the following change to that file makes a difference?

```diff
diff --git a/metakernel/magics/python_magic.py b/metakernel/magics/python_magic.py
index cb7af4c..6394c66 100644
--- a/metakernel/magics/python_magic.py
+++ b/metakernel/magics/python_magic.py
@@ -56,7 +56,7 @@ class PythonMagic(Magic):
                 self.env["__builtins__"]["raw_input"] = self.kernel.raw_input = self.kernel.raw_input
                 self.env["raw_input"] = self.kernel.raw_input = self.kernel.raw_input
         try:
-            return eval(code.strip(), self.env)
+            return eval(code.strip(), self.env, dict())
         except:
             pass
         try:
```

**Comment by kirk86** [comment]
Nope, unfortunately the same behavior.
Changed the line, restarted the notebook. Nothing :(
All the other kernels are working fine ipython, R, julia ... etc. This one is very bizarre.

**Comment by dandye** [comment]
I ran into this Issue with IPython version 2.7.8 and worked around it by modifying line 63:
```
- exec(code.strip(), self.env) 
+ exec(code.strip(), self.env) in globals(), locals()
```
Credit goes to [Lennart Regebro](http://stackoverflow.com/users/126214/lennart-regebro)'s answer on [StackOverflow In Python, why doesn't exec work in a function with a subfunction?](http://stackoverflow.com/questions/4484872/in-python-why-doesnt-exec-work-in-a-function-with-a-subfunction)

**Comment by dsblank** [comment]
Do you get the error (with the original code) if you just run:

```
ipython console --kernel=matlab_kernel
```

rather than `qtconsole`? I don't get this error using Python 2.7.6 on Ubuntu.

**Comment by kirk86** [comment]
@dsblank: I tried to run it from the console and I was receiving the same error message but fortunately @dandye's :+1: suggestion made the trick and  I can work normally. Thanks everyone. I am closing this issue for now.

**Comment by dsblank** [comment]
I've now run into this too. After updating jupyter to 4, I can't get any version of exec in this file to work in Python2.

**Comment by dsblank** [comment]
Fixed in metakernel via metakernel@605d0c8eb7c99b1ab8a40be3fdbb198166980194.
